File: services/reporting/src/reporting_mcp/tools/reports.py
"""save_report and list_reports tools — report history in reporting schema."""

import json
from typing import Any
from reporting_mcp.db import DatabaseManager


async def _save_report_impl(question: str, sql_queries: list[str], output: str, db: DatabaseManager, tags: list[str] | None = None) -> dict[str, Any]:
    rows = await db.reporting_query(
        """INSERT INTO report_history (question, sql_queries, output, tags)
           VALUES ($1, $2::jsonb, $3, $4::text[])
           RETURNING id""",
        [question, json.dumps(sql_queries), output, tags or []],
    )
    return {"report_id": rows[0]["id"]}


async def _list_reports_impl(search: str | None, db: DatabaseManager, limit: int = 20) -> list[dict]:
    if search:
        return await db.reporting_query(
            "SELECT id, question, output, tags, created_at FROM report_history WHERE question ILIKE $1 ORDER BY created_at DESC LIMIT $2",
            [f"%{search}%", limit],
        )
    return await db.reporting_query(
        "SELECT id, question, output, tags, created_at FROM report_history ORDER BY created_at DESC LIMIT $1",
        [limit],
    )


def register_report_tools(mcp, get_db):
    import json

    @mcp.tool()
    async def save_report(question: str, sql_queries: list[str], output: str, tags: list[str] | None = None) -> str:
        """Save a completed report to history for future reference."""
        return json.dumps(await _save_report_impl(question, sql_queries, output, get_db(), tags), default=str)

    @mcp.tool()
    async def list_reports(search: str | None = None, limit: int = 20) -> str:
        """Search past reports by keyword (ILIKE). Returns recent reports if no search term."""
        return json.dumps(await _list_reports_impl(search, get_db(), limit), default=str)

File: services/reporting-node/src/tools/reports.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { eq, desc, ilike, sql } from "drizzle-orm";
import type { Db } from "../db/index.js";
import { reportHistory } from "../db/schema.js";

export function registerReportTools(server: McpServer, db: Db) {
  server.tool(
    "save_report",
    "Save a completed report to history for future reference.",
    {
      question: z.string().describe("What was asked"),
      sql_queries: z.array(z.string()).describe("Queries that generated output"),
      output: z.string().describe("Result/conclusion"),
      tags: z.array(z.string()).optional().describe("Topic tags"),
    },
    async ({ question, sql_queries, output, tags }) => {
      const [inserted] = await db
        .insert(reportHistory)
        .values({
          question,
          sqlQueries: sql_queries,
          output,
          tags: tags ?? [],
        })
        .returning({ id: reportHistory.id });

      return {
        content: [
          {
            type: "text" as const,
            text: JSON.stringify({ report_id: inserted.id }),
          },
        ],
      };
    },
  );

  server.tool(
    "list_reports",
    "Search past reports by keyword (ILIKE). Returns recent reports if no search term.",
    {
      search: z.string().optional().describe("Optional keyword search"),
      limit: z.number().optional().default(20).describe("Max results"),
    },
    async ({ search, limit }) => {
      const query = db
        .select({
          id: reportHistory.id,
          question: reportHistory.question,
          output: reportHistory.output,
          tags: reportHistory.tags,
          createdAt: reportHistory.createdAt,
        })
        .from(reportHistory);

      const results = search
        ? await query
            .where(ilike(reportHistory.question, `%${search}%`))
            .orderBy(desc(reportHistory.createdAt))
            .limit(limit ?? 20)
        : await query
            .orderBy(desc(reportHistory.createdAt))
            .limit(limit ?? 20);

      return {
        content: [
          {
            type: "text" as const,
            text: JSON.stringify(results, null, 2),
          },
        ],
      };
    },
  );
}
